aztec/src/main/kotlin/org/wordpress/aztec/watchers/BlockElementWatcher.kt

```diff
@@ -19,7 +22,35 @@ open class BlockElementWatcher(aztecText: AztecText) : TextWatcher {
 
     private val aztecTextRef: WeakReference<AztecText?> = WeakReference(aztecText)
 
-    override fun beforeTextChanged(text: CharSequence, start: Int, count: Int, after: Int) { }
+    override fun beforeTextChanged(text: CharSequence, start: Int, count: Int, after: Int) {
+        if (count > 0) {
+
+            val deleteEnd = start + count
+            if (text[deleteEnd - 1] == Constants.NEWLINE && (deleteEnd - 1 == 0 || text[deleteEnd - 2] == Constants.NEWLINE)) {
+
+                // if a newline is deleted right above a heading, we want to preserve it and move it up
+                if (text[deleteEnd - 1] == Constants.NEWLINE && (deleteEnd - 1 == 0 || text[deleteEnd - 2] == Constants.NEWLINE)) {
+                    val spannable = text as Spannable
+                    val spans = SpanWrapper.getSpans(spannable, deleteEnd, deleteEnd, AztecHeadingSpan::class.java).filter { it.start == deleteEnd }
+
+                    if (spans.isNotEmpty()) {
+                        // save the text state before the funky business, then skip the history
+                        val aztecText = aztecTextRef.get()
+                        aztecText?.let {
+                            aztecText.history.beforeTextChanged(aztecText.toFormattedHtml())
+                            aztecText.consumeHistoryEvent = false
+
+                            spans.forEach {
+                                spannable.setSpan(AztecHeadingSpan(it.span.nestingLevel, it.span.TAG, it.span.attributes, (it.span as AztecHeadingSpan).headerStyle), deleteEnd - 1, deleteEnd, it.flags)
```

Comment:
I think we can remove the `as AztecHeadingSpan` cast and just use `it.span.headerStyle`

Reply:
also this line is too long? not sure if we're sticking to 130 chars max per line, but on my screen I had to scroll. It'd be more readable if we broke it like this:

```
                            spans.forEach {
                                spannable.setSpan(AztecHeadingSpan(it.span.nestingLevel, it.span.TAG, it.span.attributes,
                                        it.span.headerStyle), deleteEnd - 1, deleteEnd, it.flags)
                            }

```

Reply:
Weird, the lint checker complained about not being able to infer the type without the cast before. Maybe the new Kotlin release fixed it?